Storage/src/Connection/Rest.php

```diff
@@ -628,6 +630,167 @@ private function sendWithRetry($resource, $method, $args)
             $args
         );
 
+        $args = $this->addRetryHeaderCallbacks($args);
+
         return $this->send($resource, $method, $args);
     }
+
+    /**
+     * Adds callbacks to $args which amend retry hash and attempt to headers.
+     * @param array $args
+     *
+     * @return array
+     */
+    private function addRetryHeaderCallbacks(array $args): array
+    {
+        $requestHash = Uuid::uuid4()->toString();
+        $args['restOnRetryExceptionFunction'] = function (
+            \Exception $e,
+            $currentAttempt,
+            &$arguments
+        ) use ($requestHash) {
+            // Since we the the last attempt number here, so incrementing it
+            // to get the current attempt count
+            $this->updateRetryHeaders(
+                $arguments,
+                $requestHash,
+                $currentAttempt + 2
```

Comment:
Why is this +2 rather than +1? Is it because currentAttempt starts from zero? Might be good to add a comment to clarify. Also, please add tests for these methods; it's tricky logic and important to make sure it's accurate.

Reply:
Added the comment. Thanks for the callout :)